### PR TITLE
daemon,debian: add PolicyKit support to snapd

### PR DESCRIPTION
**Pull request by robert-ancell**
Use PolicyKit to authorize write access on the the snapd interface.

### CODE REVIEW

**Comment by snappy-m-o** [comment]
Can one of the admins verify this patch?

**Comment by robert-ancell** [comment]
The parts of this patch I'm not 100% sure about:
- CheckAuthorization() will block while a dialog is shown until the user provides a response. I don't know the snapd/go threading/loop model so I'm not sure if this will block up the snapd process or not. We can do this async with libsystemd if necessary, but again not sure how that interacts with the snapd threading model.
- I created a policy for each path - this may not be an appropriate granularity.
- We could (should?) also use policy for read actions. However this would mean if PolicyKit is not present only root works - is that fine?
- The way the credentials are converted into a string is... interesting. I hope I've done the right thing shoehorning a new variable in there.

**Comment by robert-ancell** [comment]
Also:
- The C code is just thrown in daemon.go. I'm guessing it should probably be in a separate policykit.go - please let me know if so and what form it should take.

File: daemon/api.go [hunk 1]
```diff
@@ -86,37 +86,42 @@ var (
...
 	snapsCmd = &Command{
-		Path:   "/2.0/snaps",
-		UserOK: true,
-		GET:    getSnapsInfo,
-		POST:   sideloadSnap,
+		Path:            "/2.0/snaps",
+		UserOK:          true,
+		GET:             getSnapsInfo,
+		POST:            sideloadSnap,
+		PolicyKitAction: "com.ubuntu.snappy.sideload-snap",
```
Comment (zyga) [inline daemon/api.go]: Personally I'd prefer to have a few documented constants that are referenced from all the Command objects.

File: daemon/daemon.go [hunk 2]
```diff
@@ -19,6 +19,101 @@
...
+    snprintf (filename, 1024, "/proc/%u/stat", pid);
+    f = fopen (filename, "r");
+    if (!fgets (line, 1024, f))
+        goto done;
+
+    // A line looks like this (the 22nd field is the start time - 545930 in this case):
+    // 14207 (gnome-terminal-) S 3399 3741 3741 0 -1 4194304 95756 7917395 23 857 6850 942 29913 2486 20 0 4 0 545930 692899840 12642 18446744073709551615 4194304 4486204 140735647896976 140735647896432 139643272960061 0 0 4096 65536 0 0 0 17 3 0 0 1 0 0 6585664 6596840 22155264 140735647901892 140735647901938 140735647901938 140735647903690 0
+
+    // Find the end of the name field, search from the right in case the name contains a ')'.
+    for (i = strlen (line) - 1; i > 0 && line[i] != ')'; i--);
```
Comment (zyga) [inline daemon/daemon.go]: I'd use ```i >= 0``` although here it makes little difference.

File: daemon/daemon.go [hunk 3]
```diff
@@ -19,6 +19,101 @@
...
+    uint64_t start_time = 0;
+
+    snprintf (filename, 1024, "/proc/%u/stat", pid);
+    f = fopen (filename, "r");
+    if (!fgets (line, 1024, f))
+        goto done;
+
+    // A line looks like this (the 22nd field is the start time - 545930 in this case):
+    // 14207 (gnome-terminal-) S 3399 3741 3741 0 -1 4194304 95756 7917395 23 857 6850 942 29913 2486 20 0 4 0 545930 692899840 12642 18446744073709551615 4194304 4486204 140735647896976 140735647896432 139643272960061 0 0 4096 65536 0 0 0 17 3 0 0 1 0 0 6585664 6596840 22155264 140735647901892 140735647901938 140735647901938 140735647903690 0
+
```
Comment (zyga) [inline daemon/daemon.go]: Could all of this hand-crafted manipulation be replaced by strtok_r()?

File: daemon/daemon.go [hunk 4]
```diff
@@ -79,8 +177,13 @@ func (c *Command) canAccess(r *http.Request) bool {
 		isUser = true
 	}
 
-	// only superuser can modify
+	logger.Debugf("canAccess %s %d %d '%s'", r.Method, pid, uid, c.PolicyKitAction)
```
Comment (zyga) [inline daemon/daemon.go]: "canAccess %s %d %d %q"

File: daemon/ucrednet.go [hunk 5]
```diff
@@ -28,40 +28,54 @@ import (
...
 )
 
-var errNoUID = errors.New("no uid found")
+var errNoID = errors.New("no pid/uid found")
 
+const ucrednetNoProcess = uint64(0)
 const ucrednetNobody = uint32((1 << 32) - 1)
 
-func ucrednetGetUID(remoteAddr string) (uint32, error) {
+func ucrednetGet(remoteAddr string) (uint64, uint32, error) {
```
Comment (zyga) [inline daemon/ucrednet.go]: Personally I'd replace all of the parsing below with a simple generic (and easier to verify) code that splits elements on ```';'``` and then parses each as a ```key=value```. Then the code below could just look up the key ```uid``` or ```pid```, convert it to a number and do what it has to do.

File: data/com.ubuntu.snappy.policy [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    <defaults>
+      <allow_any>auth_admin</allow_any>
+      <allow_inactive>auth_admin</allow_inactive>
+      <allow_active>auth_admin</allow_active>
+    </defaults>
+  </action>
+
+  <action id="com.ubuntu.snappy.service-snap">
+    <description gettext-domain="system-service">Service snaps</description>
+    <message gettext-domain="system-service">Authentication is required to service snaps</message>
```
Comment (zyga) [inline data/com.ubuntu.snappy.policy]: What does it mean to "service" a snap?

**Comment by zyga** [comment]
A few comments below but nothing terrible. I would love to see a check that this still builds and works fine on architectures which build go with gccgo as we had a few quirks around C interactions there.

**Comment by niemeyer** [comment]
Before reviewing this, can we have some more context for the proposal?  This is likely a good idea, but it's not yet clear if it is a good idea right now.

File: daemon/daemon.go [hunk 7]
```diff
@@ -19,6 +19,101 @@
...
+
+    result = sd_bus_message_read (reply, "bb", &authorized, &challenge);
+    if (result < 0)
+        goto done;
+
+done:
+    if (reply)
+        sd_bus_message_unref (reply);
+    if (bus)
+        sd_bus_close (bus);
```
Comment (chipaca) [inline daemon/daemon.go]: There doesn't seem to be a manpage for `sd_bus_close`, does it also do `sd_bus_unref`?

File: daemon/daemon.go (hunk 2 above)
Comment (chipaca) [inline daemon/daemon.go]: @zyga there's no valid `line` with `i < 2` anyway.

File: daemon/daemon.go (hunk 7 above)
Comment (zyga) [inline daemon/daemon.go]: There is ```sd_bus *sd_bus_unref(sd_bus *bus);```

File: daemon/daemon.go (hunk 2 above)
Comment (robert-ancell) [inline daemon/daemon.go]: It has to be i > 0 otherwise the following line will fail:
if (line[i] != ')')

File: daemon/daemon.go (hunk 3 above)
Comment (robert-ancell) [inline daemon/daemon.go]: strrchr? Yes. I think initially I was using that and it had too pointer arithmetic but it looks better now.

File: data/com.ubuntu.snappy.policy (hunk 6 above)
Comment (robert-ancell) [inline data/com.ubuntu.snappy.policy]: I don't know - I added policy for all the write operations. I don't know what service means exactly for a snap - ideas for better names welcome.

File: daemon/daemon.go (hunk 3 above)
Comment (robert-ancell) [inline daemon/daemon.go]: Fixed.

File: daemon/ucrednet.go (hunk 5 above)
Comment (robert-ancell) [inline daemon/ucrednet.go]: Fixed.

File: daemon/daemon.go (hunk 4 above)
Comment (robert-ancell) [inline daemon/daemon.go]: Fixed.

File: daemon/daemon.go (hunk 7 above)
Comment (robert-ancell) [inline daemon/daemon.go]: Fixed.

File: daemon/api.go (hunk 1 above)
Comment (robert-ancell) [inline daemon/api.go]: Fixed.

**Comment by robert-ancell** [comment]
@niemeyer the purpose of this change is to allow non-root users to install and remove snaps if the system policy allows. For example, a GUI application.

**Comment by niemeyer** [comment]
Sure, this is implicit from the code. What is the actual use case, today?

**Comment by niemeyer** [comment]
Okay, let's catch up live somewhere to discuss details of this work.

**Comment by elopio** [comment]
@robert-ancell and we need a user test for this. Talk to me or @fgimenez if you need a hand with this.

File: daemon/api.go [hunk 8]
```diff
@@ -66,6 +66,12 @@ var api = []*Command{
 	assertsFindManyCmd,
 }
 
+// PolicyKit action names
+const sideloadSnapAction = "com.ubuntu.snappy.sideload-snap"
```
Comment (sergiusens) [inline daemon/api.go]: `sideload` these days means install from somewhere different than the store and provide all the required assertions. I think you are referring to `developer` snaps here, right?

Comment (robert-ancell) [inline daemon/api.go]: I just copied the associated method name sideloadSnap() - if there's a more appropriate name then happy to switch to that.

**Comment by mvo5** [comment]
I close this pull-request temporarily because we discussed various options at the recent sprint in SC and its not clear that this branch is the one we need.

**Pull request closed without merging**